### PR TITLE
Feature/rewards-buffer

### PR DESCRIPTION
**Pull request by Web3NL**

### CODE REVIEW

File: src/canister/src/storage/reward_storage.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+use ic_cdk_timers::set_timer_interval;
+use ic_stable_structures::memory_manager::MemoryId;
+use std::{cell::RefCell, time::Duration};
+
+// Interval for sending reward activities to Reward Canister
+const INTERVAL: Duration = Duration::from_secs(24 * 60 * 60); // 1 day
+
+// timer to periodically process the reward buffer
+thread_local! {
+   pub static REWARD_TIMER: RefCell<Option<u64>> = const { RefCell::new(None) };
```
Comment (napalmpapalam) [inline src/canister/src/storage/reward_storage.rs:18]: mb lets keep storages in the one place? I read about it somewhere

File: src/canister/src/storage/reward_storage.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+use crate::logic::reward_buffer_logic::send_reward_data;
+use candid::Principal;
+use canister_types::models::reward::{Activity, RewardableActivity};
+use ic_cdk::{api::time, spawn};
+use ic_cdk_timers::set_timer_interval;
+use ic_stable_structures::memory_manager::MemoryId;
+use std::{cell::RefCell, time::Duration};
+
+// Interval for sending reward activities to Reward Canister
+const INTERVAL: Duration = Duration::from_secs(24 * 60 * 60); // 1 day
```
Comment (napalmpapalam) [inline src/canister/src/storage/reward_storage.rs:14]: mb just keep u64 coz u doing this `.as_nanos() as u64`, but use it as duration in one place where could be `Duration::from_nanos`

File: src/canister/src/storage/storage_api.rs [hunk 3]
```diff
@@ -332,9 +340,14 @@ thread_local! {
...
-        Cell::init(MEMORY_MANAGER.with(|p| p.borrow().get(HISTORY_CANISTER_MEMORY_ID)), None)
-            .expect("Failed to initialize history canister id")
+    pub static HISTORY_CANISTER: RefCell<Cell<Option<Principal>, Memory>> = RefCell::new(
+    Cell::init(MEMORY_MANAGER.with(|p| p.borrow().get(HISTORY_CANISTER_MEMORY_ID)), None)
+        .expect("Failed to initialize history canister id")
+    );
+
+    pub static REWARD_CANISTER: RefCell<Cell<Option<Principal>, Memory>> = RefCell::new(
+    Cell::init(MEMORY_MANAGER.with(|p| p.borrow().get(REWARD_CANISTER_MEMORY_ID)), None)
+        .expect("Failed to initialize reward canister id")
```
Comment (napalmpapalam) [inline src/canister/src/storage/storage_api.rs:350]:
```suggestion
    pub static HISTORY_CANISTER: RefCell<Cell<Option<Principal>, Memory>> = RefCell::new(
        Cell::init(MEMORY_MANAGER.with(|p| p.borrow().get(HISTORY_CANISTER_MEMORY_ID)), None)
            .expect("Failed to initialize history canister id")
    );

    pub static REWARD_CANISTER: RefCell<Cell<Option<Principal>, Memory>> = RefCell::new(
        Cell::init(MEMORY_MANAGER.with(|p| p.borrow().get(REWARD_CANISTER_MEMORY_ID)), None)
            .expect("Failed to initialize reward canister id")
```

Comment (napalmpapalam) [reply src/canister/src/storage/storage_api.rs:350]: not sure if it correct, but for nesting there should be 1 tab, dunno why fmt not working here

File: src/canister_types/src/models/date_range.rs [hunk 4]
```diff
@@ -3,8 +3,8 @@ use serde::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, Default, CandidType, Serialize, Deserialize)]
 pub struct DateRange {
-    start_date: u64,
-    end_date: u64,
+    pub start_date: u64,
+    pub end_date: u64,
```
Comment (napalmpapalam) [inline src/canister_types/src/models/date_range.rs:7]:
```suggestion
    pub start: u64,
    pub end: u64,
```

File: src/canister_types/src/models/reward.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    pub fn before(&self, days: u64) -> bool {
+        self.timestamp < time() - days * 24 * 60 * 60
+    }
+
+    pub fn into(self) -> RewardableActivityResponse {
+        RewardableActivityResponse {
+            activity: self.get_activity(),
+            timestamp: self.timestamp,
+        }
+    }
```
Comment (napalmpapalam) [inline src/canister_types/src/models/reward.rs:40]: implement std trait https://doc.rust-lang.org/std/convert/trait.Into.html

File: src/canister_types/src/models/reward.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    pub fn decode(bytes: Vec<u8>) -> Self {
+        Decode!(&bytes, Activity).unwrap()
+    }
+
+    pub fn as_string(&self) -> String {
+        match self {
+            Activity::GroupMemberCount(_) => "group_count".to_string(),
+            Activity::UserActivity(_) => "user_activity".to_string(),
+        }
+    }
```
Comment (napalmpapalam) [inline src/canister_types/src/models/reward.rs:69]: impl Display trait

File: src/canister_types/src/models/reward.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    pub fn new(owner: Principal, timestamp: u64) -> Self {
+        Self { owner, timestamp }
+    }
+}
+
+#[derive(CandidType, Deserialize, Clone)]
+pub struct RewardDataPackage {
+    pub group_member_counts: Vec<GroupRewardData>,
+    pub user_activity: Vec<UserActivityData>,
+}
```
Comment (napalmpapalam) [inline src/canister_types/src/models/reward.rs:105]: changelog

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -1 +1,12 @@
...
+## Reward mechanism
+
+User activity - > Reward buffer -> Periodically process buffer -> Send to reward canister
+
+User activity is logged from several places within proxy  
+Reward buffer is stored in `reward_storage.rs`  
+Timer to process buffer in `reward_buffer.rs`  
+Buffer processing logic in `reward_bufferlogic.rs`
+
+Public calls in `reward_calls.rs`
```
[the remark this answers is not in the document]
Comment (rem-code-s) [reply README.md:12]: Yes this is still on my todo list

File: src/canister/src/logic/reward_buffer_logic.rs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        match rewardable.get_activity() {
+            Activity::GroupMemberCount(id) => {
+                // collect owner, group id and member count
+                if let Ok((_, group)) = GroupStore::get(id) {
+                    let (_, members) = GroupMemberStore::get(id).unwrap_or((0, Default::default()));
+
+                    group_member_counts.push(GroupRewardData::new(
+                        group.owner,
+                        id,
+                        members.get_member_count(),
```
[the remark this answers is not in the document]
Comment (rem-code-s) [reply src/canister/src/logic/reward_buffer_logic.rs:25]: hmm there are also `invites` on the `MemberCollection` 🤔

```
pub fn get_member_count(&self) -> u64 {
        self.members.len() as u64
    }

    pub fn get_invite_count(&self) -> u64 {
        self.invites.len() as u64
    }
```

File: src/canister/src/storage/reward_storage.rs (hunk 1 above)
Comment (rem-code-s) [reply src/canister/src/storage/reward_storage.rs:18]: This storage is not persistant and gets wiped on update / reinstall

**Review by napalmpapalam — APPROVED** [review]

**Pull request merged**